### PR TITLE
Port rigid injection to the gpu

### PR DESCRIPTION
**Pull request by atmyers**

### CODE REVIEW

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 1]
```diff
@@ -90,44 +90,44 @@ RigidInjectedParticleContainer::RemapParticles()
...
 
                     const auto GetPosition = GetParticlePosition(pti);
                           auto SetPosition = SetParticlePosition(pti);
 
                     // Loop over particles
                     const long np = pti.numParticles();
-                    for (int i=0 ; i < np ; i++) {
-
+                    Real lvzbeam_ave_boosted = vzbeam_ave_boosted;
+                    Real gamma_boost = WarpX::gamma_boost;
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
                    const Real lvzbeam_ave_boosted = vzbeam_ave_boosted;
                    const Real gamma_boost = WarpX::gamma_boost;
```

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 2]
```diff
@@ -143,29 +143,38 @@ RigidInjectedParticleContainer::BoostandRemapParticles()
         vzbeam_ave_boosted = (vzbeam_ave_lab - WarpX::beta_boost*PhysConst::c)/(1. - vzbeam_ave_lab*WarpX::beta_boost/PhysConst::c);
     }
 
+    const Real csq = PhysConst::c*PhysConst::c;
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
    const Real csqi = 1./(PhysConst::c*PhysConst::c);
```

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 3]
```diff
@@ -143,29 +143,38 @@ RigidInjectedParticleContainer::BoostandRemapParticles()
...
+            const bool lfocused = focused;
+            const bool lrigid_advance = rigid_advance;
+            const Real lzinject_plane = zinject_plane;
+            amrex::ParallelFor( np, [=] AMREX_GPU_DEVICE (long i)
+            {
                 ParticleReal xp, yp, zp;
                 GetPosition(i, xp, yp, zp);
 
-                const Real gamma_lab = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])/(PhysConst::c*PhysConst::c));
+                const Real gamma_lab = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])/csq);
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
                const Real gamma_lab = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])*csqi);
```

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 4]
```diff
@@ -176,40 +185,40 @@ RigidInjectedParticleContainer::BoostandRemapParticles()
...
-                if (rigid_advance) {
+                if (lrigid_advance) {
                     // with the particle moving at the average velocity
-                    zp = zpr - vzbeam_ave_boosted*tpr;
+                    zp = zpr - lvzbeam_ave_boosted*tpr;
                 }
                 else {
                     // with the particle moving with its own velocity
-                    const Real gammapr = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])/(PhysConst::c*PhysConst::c));
+                    const Real gammapr = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])/csq);
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
                    const Real gammapr = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])*csqi);
```

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 5]
```diff
@@ -90,44 +90,44 @@ RigidInjectedParticleContainer::RemapParticles()
...
-                    for (int i=0 ; i < np ; i++) {
-
+                    Real lvzbeam_ave_boosted = vzbeam_ave_boosted;
+                    Real gamma_boost = WarpX::gamma_boost;
+                    amrex::ParallelFor( np, [=] AMREX_GPU_DEVICE (long i)
+                    {
                         ParticleReal xp, yp, zp;
                         GetPosition(i, xp, yp, zp);
 
                         const Real gammapr = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])/csq);
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
                        const Real gammapr = std::sqrt(1. + (uxp[i]*uxp[i] + uyp[i]*uyp[i] + uzp[i]*uzp[i])*csqi);
```

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 6]
```diff
@@ -90,44 +90,44 @@ RigidInjectedParticleContainer::RemapParticles()
...
                         const Real vzpr = uzp[i]/gammapr;
 
                         // Back out the value of z_lab
-                        const Real z_lab = (zp + uz_boost*t_lab + WarpX::gamma_boost*t_lab*vzpr)/(WarpX::gamma_boost + uz_boost*vzpr/csq);
+                        const Real z_lab = (zp + uz_boost*t_lab + gamma_boost*t_lab*vzpr)/(gamma_boost + uz_boost*vzpr/csq);
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
                        const Real z_lab = (zp + uz_boost*t_lab + gamma_boost*t_lab*vzpr)/(gamma_boost + uz_boost*vzpr*csqi);
```

File: Source/Particles/RigidInjectedParticleContainer.cpp [hunk 7]
```diff
@@ -90,44 +90,44 @@ RigidInjectedParticleContainer::RemapParticles()
...
 
                         // Time of the particle in the boosted frame given its position in the lab frame at t=0.
-                        const Real tpr = WarpX::gamma_boost*t_lab - uz_boost*z_lab/csq;
+                        const Real tpr = gamma_boost*t_lab - uz_boost*z_lab/csq;
```
Comment (MaxThevenet) [inline Source/Particles/RigidInjectedParticleContainer.cpp]:
```suggestion
                        const Real tpr = gamma_boost*t_lab - uz_boost*z_lab*csqi;
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 8]
```diff
@@ -646,25 +646,62 @@ std::array<Real, 3> WarpXParticleContainer::meanParticleVelocity(bool local) {
...
+            {
+                auto& ux = pti.GetAttribs(PIdx::ux);
+                auto& uy = pti.GetAttribs(PIdx::uy);
+                auto& uz = pti.GetAttribs(PIdx::uz);
+
+                np_total += pti.numParticles();
+
+                for (unsigned long i = 0; i < ux.size(); i++) {
+                    Real usq = (ux[i]*ux[i] + uy[i]*uy[i] + uz[i]*uz[i])*inv_clight_sq;
+                    Real gaminv = 1.0/std::sqrt(1.0 + usq);
```
Comment (MaxThevenet) [inline Source/Particles/WarpXParticleContainer.cpp]:
```suggestion
                    Real gaminv = 1.0_rt/std::sqrt(1.0_rt + usq);
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 9]
```diff
@@ -646,25 +646,62 @@ std::array<Real, 3> WarpXParticleContainer::meanParticleVelocity(bool local) {
...
+                long np = pti.numParticles();
+                np_total += np;
+
+                reduce_op.eval(np, reduce_data,
+                               [=] AMREX_GPU_DEVICE (int i) -> ReduceTuple
+                               {
+                                   Real usq = (uxp[i]*uxp[i] +
+                                               uyp[i]*uyp[i] +
+                                               uzp[i]*uzp[i])*inv_clight_sq;
+                                   Real gaminv = 1.0/std::sqrt(1.0 + usq);
```
Comment (MaxThevenet) [inline Source/Particles/WarpXParticleContainer.cpp]:
```suggestion
                                   Real gaminv = 1.0_rt/std::sqrt(1.0_rt + usq);
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 10]
```diff
@@ -646,25 +646,62 @@ std::array<Real, 3> WarpXParticleContainer::meanParticleVelocity(bool local) {
...
+        ReduceData<Real, Real, Real> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+        for (int lev = 0; lev <= nLevels; ++lev) {
+            for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti)
+            {
+                const auto uxp = pti.GetAttribs(PIdx::ux).data();
+                const auto uyp = pti.GetAttribs(PIdx::uy).data();
+                const auto uzp = pti.GetAttribs(PIdx::uz).data();
+
+                long np = pti.numParticles();
```
Comment (MaxThevenet) [inline Source/Particles/WarpXParticleContainer.cpp]:
```suggestion
                const long np = pti.numParticles();
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 11]
```diff
@@ -646,25 +646,62 @@ std::array<Real, 3> WarpXParticleContainer::meanParticleVelocity(bool local) {
...
     const int nLevels = finestLevel();
-    for (int lev = 0; lev <= nLevels; ++lev) {
 
+#ifdef AMREX_USE_GPU
+    if (Gpu::inLaunchRegion())
+    {
+        ReduceOps<ReduceOpSum, ReduceOpSum, ReduceOpSum> reduce_op;
+        ReduceData<Real, Real, Real> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+        for (int lev = 0; lev <= nLevels; ++lev) {
```
Comment (MaxThevenet) [inline Source/Particles/WarpXParticleContainer.cpp]: Could you explain a little bit what this is doing?

Comment (atmyers) [reply Source/Particles/WarpXParticleContainer.cpp]: This is how we can do reductions on the GPU - in particular, using the `ReduceTuple`, we can compute the total `vx`, `vy`, and `vz` in one kernel launch.

**Comment by atmyers** [comment]
Seems like the "initial distribution" test is failing on the CI. Do I need to merge with master?

**Comment by MaxThevenet** [comment]
@Yin-YinjianZhao could you try and answer @atmyers's question above about the initial distribution test?

**Comment by MaxThevenet** [comment]
@atmyers the `initial_distribution` test passed on master and I din't think it's been changed. However, it relies on rigid injection. Is it possible that we get division by zero in the python script because no particle is injected in the simulation due to rigid injection issue?

File: Source/Particles/WarpXParticleContainer.cpp [hunk 12]
```diff
@@ -646,25 +646,62 @@ std::array<Real, 3> WarpXParticleContainer::meanParticleVelocity(bool local) {
...
+            {
+                auto& ux = pti.GetAttribs(PIdx::ux);
+                auto& uy = pti.GetAttribs(PIdx::uy);
+                auto& uz = pti.GetAttribs(PIdx::uz);
+
+                np_total += pti.numParticles();
+
+                for (unsigned long i = 0; i < ux.size(); i++) {
+                    Real usq = (ux[i]*ux[i] + uy[i]*uy[i] + uz[i]*uz[i])*inv_clight_sq;
+                    Real gaminv = 1.0_rt/std::sqrt(1.0_rt + usq);
```
Comment (Yin-YinjianZhao) [inline Source/Particles/WarpXParticleContainer.cpp]:
```suggestion
                    Real gaminv = std::sqrt(1.0_rt + usq);
```

Comment (Yin-YinjianZhao) [reply Source/Particles/WarpXParticleContainer.cpp]: Never mind, didn't notice `gaminv` means gamma inverse.

File: Source/Particles/WarpXParticleContainer.cpp (hunk 11 above)
Comment (MaxThevenet) [inline Source/Particles/WarpXParticleContainer.cpp]: I think the name `nLevels` isn't correct, this is `max_level`, if you agree I can make a separate PR to fix this throughout `PhysicalParticleContainer.cpp`.

Comment (MaxThevenet) [reply Source/Particles/WarpXParticleContainer.cpp]: Nice feature! Thanks for the explanation.

Comment (atmyers) [reply Source/Particles/WarpXParticleContainer.cpp]: Yeah you are right.

**Pull request merged**